lib/puppet/configurer.rb

```diff
@@ -516,7 +516,7 @@ def retrieve_catalog_from_cache(query_options)
     return nil
   end
 
-  def retrieve_new_catalog(query_options)
+  def retrieve_new_catalog(facts, query_options)
```

Comment:
Still need to call the httpclient directly when terminus is `:rest`

Reply:
We can't call the httpclient directly, because there's no guarantee that the `terminus_class == :rest` for the agent, apply and device applications. We could special case that, but just call the catalog find method for now, passing the facts object.